http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/sender/RedirectHandler.java

```diff
@@ -0,0 +1,782 @@
+        }
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("Channel " + ctx.channel().id() + " gets inactive so closing it from RedirectHandler.");
+        }
+        if (originalChannelContext == ctx) { //If this is the original channel it must be destroyed from the pool
+            ctx.fireChannelInactive();
+        } else {
+            ctx.close();
+        }
+    }
+
+    /**
+     * Handle redirect state for the channel.
+     *
+     * @param ctx ChannelHandler context
+     * @param msg Response message
+     */
+    private void handleRedirectState(ChannelHandlerContext ctx, HttpResponse msg) throws Exception {
+        try {
+            originalRequest = ctx.channel().attr(Constants.ORIGINAL_REQUEST).get();
+            String location = getLocationFromResponseHeader(msg);
+            int statusCode = msg.status().code();
+            if (location != null) {
+                redirectState = getRedirectState(location, statusCode, originalRequest);
+            } else {
+                redirectState = null;
+            }
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("Handling redirect state for channel : " + ctx.channel().id());
+            }
+            if (isRedirectEligible()) {
+                isCrossDoamin = isCrossDomain(location, originalRequest);
+                currentRedirectCount = updateAndGetRedirectCount(ctx);
+                if (currentRedirectCount <= maxRedirectCount) {
+                    if (LOG.isDebugEnabled()) {
+                        LOG.debug("Redirection required.");
+                    }
+                    isRedirect = true;
+                } else {
+                    if (LOG.isDebugEnabled()) {
+                        LOG.debug("Maximum redirect count reached.");
+                    }
+                    isRedirect = false;
+                    sendResponseHeadersToClient(ctx, msg);
+                }
+            } else {
+                if (LOG.isDebugEnabled()) {
+                    LOG.debug("Request is not eligible for redirection.");
+                }
+                isRedirect = false;
+                if (ctx == originalChannelContext) {
+                    originalChannelContext.fireChannelRead(msg);
+                } else {
+                    sendResponseHeadersToClient(ctx, msg);
+                }
+            }
+        } catch (UnsupportedEncodingException exception) {
+            LOG.error("UnsupportedEncodingException occurred when deciding whether a redirection is required",
+                    exception);
+            exceptionCaught(ctx, exception.getCause());
+        } catch (MalformedURLException exception) {
+            LOG.error("MalformedURLException occurred when deciding whether a redirection is required", exception);
+            exceptionCaught(ctx, exception.getCause());
```

Comment:
Are we logging the exception twice? Shall we log it only once? Maybe we can pass the exception message as well to the exception caught method for logging

Reply:
This is fixed now. Since the exceptionCaught is an overriden method from netty framework, to handle application logic errors, a separate method was defined. https://github.com/wso2/carbon-transports/pull/415